Priority levels (#390)
rust-lang/rust-forge: src/compiler/prioritization/priority-levels.md

# Priority levels

As the compiler team's resources are limited, the prioritization working group's main goal is to identify the most relevant issues to work on, so that the compiler team can focus on what matters the most.

## Words used in this document:

> `issue` refers to bugs and feature requests that are nominated for prioritization, by flagging the `I-Prioritize` label as described below.

Reviewer: ```suggestion
`issue` refers to bugs and feature requests that are nominated for prioritization, by flagging the `I-prioritize` label as described below.
```

This document will define what each label means, and what strategy for each label will be used.

## Labels

> Labeling an issue as `I-Prioritize` starts the prioritization process, which will end by removing the `I-Prioritize` label and appending one of the 4 labels we will discuss below:

Reviewer: ```suggestion
Labeling an issue as `I-prioritize` starts the prioritization process, which will end by removing the `I-prioritize` label and appending one of the 4 labels we will discuss below:
```

They can be `P-critical` issues that have a mitigating condition as defined above, or important issues that aren't deemed blockers.

Because there are too many `P-high` issues to fit in every compiler meeting, they should rather be handled asynchronously by the Prioritization WG, in order to help them move forward. They can still occasionally be brought up at meetings when it is deemed necessary.

The effectiveness of the Prioritization WG will be a direct consequence of our ability to draw the line between `P-critical` and `P-high` issues. There shouldn't be too many `P-critical` issues that compiler meetings become unmanageable, but critical issues shouldn't get lost in the list of P-high issues.

P-high issues are issues the teams will mostly work on. We want to make sure they're assigned, and keep an eye on them.

## P-medium and P-low

`P-medium` refer to issues that aren't a priority for the team, and that will be resolved in the long run. Eg issues that will be fixed after a specific feature has landed.

`P-low` refer to issues issue that the compiler team doesn't plan to resolve, but are still worth fixing.

> P-medium issues won't be our focus. 

Reviewer: ```suggestion
`P-medium` issues won't be our focus. 
```

They are issues we would mentor someone interested in fixing. 
They will remain in this state until someone complains, a community member fixes it, or it gets fixed by accident.

> ---

Reviewer: Do the formatting starting here make sense?. I'm not sure what's the intention, not saying that is bad :smile: just wondering if these are leftovers

> ###### For questions and comments about this document feel free to hit the relevant [zulip topic](https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/What.20is.20the.20meaning.20of.20each.20priority.20level.3F) or [open an issue](https://github.com/rust-lang/rust-forge/issues).

Reviewer: If you consider leaving this, is the best to leave it as a title?

Author: It's probably not useful anymore, people can submit issues / prs here instead :)

> ######

Reviewer: Mainly this :point_up:

# Priority levels

As the compiler team's resources are limited, the prioritization working group's main goal is to identify the most relevant issues to work on, so that the compiler team can focus on what matters the most.

## Words used in this document:

`issue` refers to bugs and feature requests that are nominated for prioritization, by flagging the `I-Prioritize` label as described below.

This document will define what each label means, and what strategy for each label will be used.

## Labels

Labeling an issue as `I-Prioritize` starts the prioritization process, which will end by removing the `I-Prioritize` label and appending one of the 4 labels we will discuss below:

> - P-Critical

Reviewer: ```suggestion
- P-critical
```

> - P-High

Reviewer: ```suggestion
- P-high
```

> - P-Medium

Reviewer: ```suggestion
- P-medium
```

> - P-Low

Reviewer: ```suggestion
- P-low
```

Each of these labels defines a strategy the team will adopt regarding:

- The amount of focus a given issue will receive
- How members of the community can get involved

## Definitions

> ### P-Critical

Reviewer: ```suggestion
### P-critical
```

Reviewer 2: @spastorino why `P-critical` but `P-Medium` (see line 66)?

> A `P-Critical` issue is a potentially blocker issue.

Reviewer: ```suggestion
A `P-critical` issue is a potentially blocker issue.
```

    - If the code in question is theoretical and considered unlikely to exist in the wild, or if it only exists in small, unmaintained packages that are not widely used
  - If a regression has been in stable for a release or two (either because we are still awaiting a fix, or because the bug had laid dormant i.e. undetected), we typically lower the priority as well, because by that time, if the users have not raised a ruckus about the regression, that is a sign that it is inherently not a critical issue. Eg: [an issue that would have been P-Critical but ended up being P-High](https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/pre-meeting.20triage.202020-04-09.20.2354818)

- Regressions where code still compiles but does something different than it used to do (dynamic semantics have changed)
  - Mitigating conditions that may lower priority:
    - If code uses feature that is explicitly not specified (e.g. `std::vec::Vec` docs state order in which it drops its elements is subject to change)
- Feature-gated features accessible without a feature gate
  - Mitigating conditions that may lower priority:
    - If the pattern is VERY unlikely

> - Soundness holes where common code that should not compile actually does

Reviewer 3: I think miscompilations are covered by the first bullet, "Regressions where code still compiles but does something different than it used to do (dynamic semantics have changed)", ...

... though arguably that is "solely" talking about regressions. It might be better to talk about deviations from the *intended semantics* (rather than the historical semantics)

  - Mitigating conditions that may lower priority:
    - Soundness holes that are difficult to trigger
    - Soundness holes that have been around for a very long time may be critical, but typically require

> - Diagnostic regressions where the diagnostic is very common and the situation very confusing

Author: what is our stance on that  ?

> - Diagnostic regressions where the diagnostic is very common and the situation very confusing

Reviewer: I guess we can just leave it there, would be useful as a reminder. Of course the phrase is correct in some sense, if the compilation error is completely misleading for a very common thing, I guess the issue is `P-critical`, imagine a move error being reported as a something completely misleading.

- ICEs for common scenarios or code patterns
  - Mitigating conditions that may lower priority:
    - If the code that triggers the ICE also triggers compilation errors, and those errors are emitted before the ICE

>     - If the code in question makes use of unstable features, particularly if the ICE requires a feature gate

Author: I m not really sure I understand what I'm supposed to write here, can you please elaborate a bit ?

### P-High

> `P-High` issues are the ones the compiler team will focus on the most.

Reviewer: `P-critical` are the ones that the compiler team focuses the most.

Maybe just remove this and start with `P-high` issues can be `P-critical` issues that have mitigating ...

Reviewer 4: I think we can keep it, but start with a base definition and rearrange the wording. Suggestion:

### P-high

`P-high` issues are issues that need attention from the compiler team, but not to the point that they need to be discussed at every meeting. They can be `P-critical` issues that have a mitigating condition as defined above, or important issues that aren't deemed blockers.

Because there are too many `P-high` issues to fit in every compiler meeting, they should rather be handled *asynchronously* by the Prioritization WG, in order to help them move forward. They can still occasionally be brought up at meetings when it is deemed necessary.

The effectiveness of the Prioritization WG will be a direct consequence of our ability to draw the line between `P-critical` and `P-high` issues. There shouldn't be too many `P-critical` issues that compiler meetings become unmanageable, but critical issues shouldn't get lost in the list of `P-high` issues.

They can be `P-Critical` issues that have a mitigating condition as defined above, or important issues that aren't deemed blockers.

The effectiveness of the prioritization Working Group will be a direct consequence of our ability to draw the line between `P-Critical` and `P-High` issues.

### P-Medium

> `P-Medium` refer to issues that aren't critical, and that will be resolved in the long run. Eg issues that will be fixed after a specific feature has landed.

Reviewer: ```suggestion
`P-Medium` refer to issues that aren't a priority for the team, and that will be resolved in the long run. Eg issues that will be fixed after a specific feature has landed.
```

### P-Low

> `P-Low` bugs are bugs that don't cause soundness or safety issues, and that occur in a "failing path" already. Eg: Code that wouldn't compile and raises an error. It might also refer to features that would be nice to have, but aren't deemed a priority

Reviewer: I'm not loving the definition of P-low we have here. I think we've documented the different levels in a different place elsewhere? maybe we can reuse some explanations from other documents. Don't remember to be honest what we have. cc @rust-lang/wg-prioritization

Reviewer 4: I might have seen a definition like "issue that the compiler team doesn't plan to resolve, but are still worth fixing".

## Amount of focus a given issue will receive

### P-Critical

> A P-critical issue will receive the most attention. It must be assigned one of several people as soon as possible, and the rest of the team should do their best to help them out if/when applicable.

Reviewer 4: Typo?
```suggestion
A P-critical issue will receive the most attention. It must be assigned one or several people as soon as possible, and the rest of the team should do their best to help them out if/when applicable.
```

### P-High

P-High issues are issues the teams will mostly work on. We want to make sure they're assigned, and keep an eye on them.

### P-Medium

P-Medium issues won't be our focus. They are issues we would mentor someone interested in fixing.

### P-Low

> P-Low issues won't be our focus. They will remain in this state until someone complains, a community member fixes it, or it gets fixed by accident.

Reviewer 4: We can probably regroup `P-medium` and `P-low` issues in this category?
